src/core/server/http/http_service.ts

```diff
@@ -72,10 +74,60 @@ export class HttpService
       configService.atPath<ExternalUrlConfigType>(externalUrlConfig.path),
     ]).pipe(map(([http, csp, externalUrl]) => new HttpConfig(http, csp, externalUrl)));
     const shutdownTimeout$ = this.config$.pipe(map(({ shutdownTimeout }) => shutdownTimeout));
+    this.prebootServer = new HttpServer(logger, 'Preboot', shutdownTimeout$);
     this.httpServer = new HttpServer(logger, 'Kibana', shutdownTimeout$);
     this.httpsRedirectServer = new HttpsRedirectServer(logger.get('http', 'redirect', 'server'));
   }
 
+  public async preboot(deps: PrebootDeps): Promise<InternalHttpServicePreboot> {
+    this.log.debug('setting up preboot server');
+    const config = await this.config$.pipe(first()).toPromise();
+
+    const prebootSetup = await this.prebootServer.setup(config);
+    prebootSetup.server.route({
+      path: '/{p*}',
+      method: '*',
+      handler: (req, responseToolkit) => {
```

Comment:
The route here covers all HTTP methods, and the one in `CoreApp` is just for `GET` (resources), but it's a bit tricky. I decided to keep this one here to preserve original behavior since we don't have `preboot` plugins yet (and also we don't register `preboot` resources routes in this case).

Once we have a `userSetup` preboot plugin we'll have to decide how to respond to requests to the "standard" API endpoints when Kibana is at the `preboot` stage. There can be some automation built around Kibana, so we cannot just reply with `404` as it would be confusing and logically not correct: endpoint exists, but not at this stage and consumers don't have to be aware of these stages, 503 would still make more sense.

For `GET` the handling logic will probably be a bit more complex: we may want to still reply with `503` to requests with `kbn-xsrf` and render anonymous app otherwise (assuming setup IS on hold, but if not - good old 503).

In any case, I feel like it'd be better to tackle this separately in a follow-up where we'll be enabling `userSetup` preboot plugin.

Reply:
Sounds good to me, thanks for the clarification